lib/scholar/decomposition/incremental_pca.ex

```diff
@@ -0,0 +1,355 @@
+      """
+    ]
+  ]
+
+  tensor_opts =
+    stream_opts ++
+      [
+        batch_size: [
+          type: :pos_integer,
+          doc: "The number of samples in a batch."
+        ]
+      ]
+
+  @stream_schema NimbleOptions.new!(stream_opts)
+  @tensor_schema NimbleOptions.new!(tensor_opts)
+
+  @doc """
+  Fits an Incremental PCA model.
+
+  ## Options
+
+  #{NimbleOptions.docs(@tensor_schema)}
+  """
+  deftransform fit(%Nx.Tensor{} = x, opts) do
+    if Nx.rank(x) != 2 do
+      raise ArgumentError,
+            "expected input tensor to have shape {num_samples, num_features},
+             got tensor with shape: #{inspect(Nx.shape(x))}"
+    end
+
+    opts = NimbleOptions.validate!(opts, @tensor_schema)
+
+    {num_samples, num_features} = Nx.shape(x)
+
+    batch_size =
+      if opts[:batch_size] do
+        opts[:batch_size]
+      else
+        5 * num_features
+      end
+
+    num_components = opts[:num_components]
+
+    cond do
+      num_components > batch_size ->
+        raise ArgumentError,
+              """
+              num_components must be less than or equal to \
+              batch_size = #{batch_size}, got #{num_components}
+              """
+
+      num_components > num_samples ->
+        raise ArgumentError,
+              """
+              num_components must be less than or equal to \
+              num_samples = #{num_samples}, got #{num_components}
+              """
+
+      num_components > num_features ->
+        raise ArgumentError,
+              """
+              num_components must be less than or equal to \
+              num_features = #{num_features}, got #{num_components}
+              """
+
+      true ->
+        nil
+    end
+
+    opts = Keyword.put(opts, :batch_size, batch_size)
+
+    fit_n(x, opts)
+  end
+
+  defn fit_n(x, opts) do
+    batch_size = opts[:batch_size]
+
+    {batches, leftover} =
+      get_batches(x, batch_size: batch_size, min_batch_size: opts[:num_components])
+
+    num_batches = Nx.axis_size(batches, 0)
+    model = fit_first_n(batches[0], opts)
+
+    {model, _} =
+      while {
+              model,
+              {
+                batches,
+                i = Nx.u64(1)
+              }
+            },
+            i < num_batches do
+        batch = batches[i]
+        model = partial_fit_n(model, batch)
+        {model, {batches, i + 1}}
+      end
+
+    model =
+      case leftover do
+        nil -> model
+        _ -> partial_fit_n(model, leftover)
+      end
+
+    model
+  end
+
+  @doc """
+  Fits an Incremental PCA model on a stream of batches.
+
+  ## Options
+
+  #{NimbleOptions.docs(@stream_schema)}
+  """
+  def fit(batches = %Stream{}, opts) do
+    opts = NimbleOptions.validate!(opts, @stream_schema)
+    first_batch = Enum.at(batches, 0)
+    model = fit_first_n(first_batch, opts)
+    batches = Stream.drop(batches, 1)
```

Comment:
This will iterate some lazy streams (such as file) twice. For eample, we will open it once to get the first batch and then again to traverse it. The correct way is to have a single reduce, which it starts with an accumulator of `{:head, ...}` and then you swap it to `{:tail, ...}` for the remaining iterations.

Although I wonder if we should have this function here... perhaps it is best to have `fit_head` and `fit_tail` as public functions and then later on we provide facilities for streaming on any model that has `fit_head` and `fit_tail`?

Reply:
> One problem with doing this using a single function is that there needs to be some starting model as initial accumulator.

This can probably be achieved by setting `num_samples_seen` to `0` and filling tensor fields with `nan`.

Reply:
There is no need to have an initial accumulator, you could have the acc be `:first` and then `model`, and pattern matching distinguighes between them.

But you could also have a model with a single `:initialized` field as a boolean (or even a u8) and then we could easily fold it all into a single `defn`.

Reply:
> There is no need to have an initial accumulator, you could have the acc be :first and then model, and pattern matching distinguighes between them.

I see, that would work.

> But you could also have a model with a single :initialized field as a boolean (or even a u8) and then we could easily fold it all into a single defn.

Checking if `num_samples_seen` is equal to `0` can tell us if the model has been initialised or not. The only issue would be the floating point type of tensor fields, but then we can ask the user to provide it.

Reply:
Alright, it is possible to have a single reduce, but then the problem is that the user needs to specify the type and the number of features as those cannot be inferred without running the stream.

I suppose it is a cleaner approach to have a reducer with two clauses: one for head and one for tail.